RUN-2988: Add job reference to the converted workflow tab

smartinellibenedetti opened the pull request:
<!--
IMPORTANT: Before submitting your Pull Request, please review the following instructions:

1. Please follow the [Developer Guidelines](https://github.com/rundeck/rundeck/wiki/Developer-Guidelines) document.
2. Are you implementing a feature or enhancement? Please search the existing Issues and look 
   at the [Rundeck Roadmap Trello Board](https://trello.com/b/sn3g9nOr/rundeck-development) for your idea before posting.
   If your enhancement is not listed, it is better to 
   [post a new enhancement request](https://github.com/rundeck/rundeck/issues/new?template=feature_request.md)
   and get feedback from maintainers and the community *before* submitting an Pull request to implement it.
3. Please be sure the issue you are addressing is referenced in a commit, or the body of your PR,
   using [github keywords](https://help.github.com/articles/closing-issues-using-keywords/), e.g. "fixes #123".
-->

**Is this a bugfix, or an enhancement? Please describe.**
<!-- Include a clear and concise description of the bug you are fixing (reference issue number), or enhancement you are implementing. -->

Convert the job reference step to Vue and add it to the already converted workflow tab components.

**Describe the solution you've implemented**
<!-- A clear and concise description of what you want to happen. -->

As the job reference step isn't an actual plugin, hardcoded it as a new form component and as a plugin option whenever fetching WorkflowNodeSteps and WorkflowSteps.

The form component connects to the node-filter-input and to a newly created pinia store, that fetches and filters nodes. Connects to the job-list-page, to display the jobs available within the selected project. And updates the later component to use buttons instead of links depending on the props passed, besides emitting via eventBus, when there are no jobs available to display.

Also fix an issue with ui-socket, that was preventing the visible filter from being correctly applied.

**Describe alternatives you've considered**
<!-- A clear and concise description of any alternative solutions or features you've considered. -->

A few alternatives:
- Originally, the goal of this ticket was also to consume a different API for fetching nodes, but unfortunately, that API was missing attributes (isLocal, and attributes) that would require changing how the resource format works for nodes, or doing big changes within node-list-embed and its children to handle it;
- Considered to add a new flag to EditPluginModal to disable fetching the provider, in order to have the logic of job reference step inside of it. Discarded due to concern of how big the component would be;
- Considered to try to format the form for job reference in a way that would be compatible with plugin-config. Discarded due to complication of integrating with the node search bar;
- Considered to not create the NodeStorePinia, calling instead the service without storing data and leveraging the existing nodeFilterStore. Discarded due to the intention that the pinia store will make it easier in the long run to update other parts of the project that rely on it;

Also:
- Considered to not have the hardcoded logic of the job ref inside the plugins store. Discarded to avoid sorting the data twice. Open to other suggestions;

**Additional context**
<!-- Add any other context or screenshots about the change here. -->

gschueler commented on rundeckapp/grails-app/views/scheduledExecution/edit.gsp:
```diff
@@ -6,6 +6,10 @@
     <meta name="tabpage" content="jobs"/>
     <meta name="layout" content="base"/>
     <asset:javascript src="static/pages/dynamic-form.js" defer="defer"/>
+    <asset:javascript src="static/pages/job/browse.js" defer="defer"/>
+    <asset:stylesheet href="static/css/pages/job/browse.css" />
+
+    <g:embedJSON id="jobTreeUiMeta" data="[hideActions: true, hideHeader: true]"/>
```
this is missing from `create.gsp`.  however you could just move it to `_jobEditCommon.gsp` instead

gschueler replied:
Job browser doesn't work on "Create" page

gschueler commented on rundeckapp/grails-spa/packages/ui-trellis/src/app/components/job/workflow/JobRefForm.vue:
```diff
@@ -0,0 +1,820 @@
+      section="job-list-page"
+      location="card-header"
+    />
+    <ui-socket
+      section="job-list-page"
+      location="tree-browser"
+      :socket-data="{
+        showBulkEdit: false,
+        showCreateButton: false,
+        projectToDisplay: editModel.jobref.project,
+        key: editModel.jobref.project,
+      }"
+    />
```
question: why not use the job-list-page component directly?

smartinellibenedetti replied:
That was the intention, but the moment I've used it, I realized that it would require parts of the logic that currently live in the main.ts. I didn't want to duplicate code, but also didn't have bandwidth to extract that logic into a new component, therefore opted to keep using the ui-socket.

gschueler reviewed (changes requested):
The Job Browser selector has another issue: it shows the Folder icon for groups, and if you click it, it changes the URL without navigating away, and then focuses on the selected group without a way to navigate back.  It should be disabled, or that behavior fixed.

smartinellibenedetti commented:
@gschueler Fixes done:

- disabled folder navigation in the browserGroupItem when allowFolderNavigation is false;
- fixed issue where the job selection wasn't working during job creation (was missing import);
- added a message when the number of results for nodes is bigger than 20;
- node overrides section will be loaded open when editing a step that has this info set (either a filter OR set nodeIntersection value, which matches original behavior);
- added truncation to the list of arguments (https://github.com/rundeck/rundeck/issues/9440);
- improved labels for finding the job via name or uuid (moved the description to inside the label);
- fixed edge case for search: if the user erases the filter previously set, the UI will show back the message that the nodes originally set by the job will be used;

Also, about this comment:

> It seems we have to turn on the "rundeck.feature.alphaUi.enabled=true" to try the new Workflow Editor
> 
> However, do we also have to use nextUi=true in the URL? The checkbox in the Cog menu has been removed to set the Cookie.
> 
> If nextUi is not set, then the Workflow Editor is not shown.
> 
> Originally, I wanted to allow the "next" UI changes to be tested by users with the checkbox. (such as the Options editor). Perhaps we can move Options out from behind the flag. In that case, maybe "alphaUi" should be used only for the Workflow Editor.

I had adjusted in #9535 to have everything behind the nextUI + alpha - both to be able to consolidate the info and also because I didn't see this comment before. I'd like to ask if we can have the workflow tab just behind the nextUI flag, as it is so close to the conclusion.

gschueler reviewed (commented):
request: the only change I think would be good is to make the modal for selecting step plugin/job reference wider.  Even with a wide screen, the modal is very narrow
